torchdynamo/eval_frame.py

```diff
@@ -79,6 +87,28 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         self.backend_ctx.__exit__(exc_type, exc_val, exc_tb)
 
     def __call__(self, fn):
+        fn = find_unaltered_fn(fn)
+        # Optimize the forward method of torch.nn.Module object
+        if isinstance(fn, torch.nn.Module):
+            mod = fn
+            optimized_forward = self(mod.forward)
+
+            class Wrapper:
+                def __getattr__(self, name):
+                    return getattr(mod, name)
+
+                def forward(self, *args, **kwargs):
+                    return optimized_forward(*args, **kwargs)
+
+                def __call__(self, *args, **kwargs):
+                    return self.forward(*args, **kwargs)
+
+            new_mod = Wrapper()
+            # Save the function pointer to find the original callable while nesting
+            # of decorators.
+            new_mod._torchdynamo_orig_callable = mod
+            return new_mod
```

Comment:
I don't think this is necessary. If you are passed a bound method (e.g., `module.forward`), it is still associated with the original Module object, so you can wrap it and it will still end up operating on the original module. This is what Dmytro was doing in the VHD model, btw

Reply:
This is for the case like this

`optimized_mod = torchdynamo.optimize("eager")(mod)`

In this case, we just want to optimized the `forward` method and for everything else, we want to re-direct to original module.

Do you mean something else? I am gonna look in VHD model to be sure.

Reply:
I don't think users should expect `optimize` to work this way. The thing expects a callable, you give a callable back, there isn't any extra structure you should expect.

I suppose we could have a `torchdynamo.optimize_module` in analogy to AOTAutograd's module-level equivalent, but I don't really see the point; dynamo has no need to see the module, it can figure it all out even if the module is a free variable to the function in question.

Reply:
I see. I am ok with that. The question then is - what does UX looks like?

For functions, its good
```
@torchdynamo.optimize("eager")
def fn():
   .....

or
optimized_fn = torchdynamo.optimize("eager")(fn)
```

What about nn.Modules? The following seems sketchy (mutating the original mod).

```
mod.forward = torchdynamo.optimize("eager")(mod.forward)
```

Or do we say that wrap the mod call in another function and then optimize that function? Something like

```
@torchdynamo.optimize("eager")
def fn(....):
    return mod(...)
```

Reply:
`mod` here is callable too, you would just have to trace it as `mod(args)` and not `mod.forward(args)`. That's not the same, because iirc while the module callable does call forward, it also does more than that with hooks and such. 

> `mod.forward = torchdynamo.optimize("eager")(mod.forward)`

I also wanted to ask, as in both your cases you are taking out the `optimized_fn ` to use beyond the Dynamo context. When would you need to do that? Generally, if you want to take the result of `.optimize`, you would use `.export`

`mod.forward = torchdynamo.export(mod.forward, args)`

Which should work fine, I believe.

Reply:
Oh, I misunderstood, I think, you are not calling to trace, you are calling to store a dynamo wrapped function.

In that case, what is wrong with something like 

```
@torchdynamo.optimize("eager")
def fn(....):
    return mod(...)
``` ?

Reply:
Yes, I am calling to store a dynamo wrapper function. It will get traced later on the first invocation. Background is that we are trying to remove the TorchDynamo context manager, because it confuses users.

I am totally ok with the following wrapping. TorchDynamo expects a callable, and returns a callable. Simple and clear.

```
@torchdynamo.optimize("eager")
def fn(....):
    return mod(...)
```

Here, the expectation from user is to encapsulate the code that they want to be optimized in another function, and then call `torchdynamo.optimize` on that function. So, just confirming, if that's ok.

Reply:
cc @jansel

Reply:
I just want to point out that the wordy version here is equivalent to `torchdynamo.optimize("eager")(mod)` (or, perhaps the eta expansion matters, in which case it's equivalent to `torchdynamo.optimize("eager")(lambda *args, **kwargs: mod(*args, **kwargs))`).

Reply:
However, I have also noticed that this behaves differently to using the context manager, so it feels like there is some sort of latent bug somewhere around here. I have a repro in fbcode if that would help.

Reply:
The main argument I see forwarding `nn.Module()` methods is that is what `fx.symbolic_trace()` and `torch.jit.script()` do, so users of those systems might expect it.